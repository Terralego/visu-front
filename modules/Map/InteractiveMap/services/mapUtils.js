export function toggleLayerVisibility (map, layerId, visibility) {
  map.setLayoutProperty(layerId, 'visibility', visibility);
}

export function getOpacityProperty (type) {
  switch (type) {
    case 'background':
      return 'background-opacity';
    case 'fill':
      return 'fill-opacity';
    case 'line':
      return 'line-opacity';
    case 'symbol':
      return 'icon-opacity';
    case 'raster':
      return 'raster-opacity';
    case 'circle':
      return 'circle-opacity';
    case 'fill-extrusion':
      return 'fill-extrusion-opacity';
    default:
      return null;
  }
}

export function setLayerOpacity (map, layerId, opacity) {
  const layer = map.getLayer(layerId);
  const property = getOpacityProperty(layer.type);
  if (property) {
    map.setPaintProperty(layerId, property, opacity);
  }
}

export function getInteractionOnEvent ({ eventType, map, point, interactions }) {
  const features = map.queryRenderedFeatures(point);
  let interaction = false;

  features.some(feature => {
    const { layer: { id: layerId } } = feature;

    const foundInteraction = interactions.find(({ id, trigger = 'click' }) => {
      const found = id === layerId;

      return found &&
        eventType === (trigger === 'mouseover' ? 'mousemove' : trigger);
    });

    if (!foundInteraction) return false;

    interaction = {
      interaction: foundInteraction,
      feature,
      layerId,
    };

    return true;
  });

  return interaction;
}

export function setInteractions ({ map, interactions, callback }) {
  const eventsTypes = new Set(interactions.reduce((triggers, { trigger = 'click' }) => [...triggers, trigger], []));

  if (eventsTypes.has('mouseover')) {
    eventsTypes.add('mousemove');
    eventsTypes.delete('mouseover');
  }

  eventsTypes.forEach(eventType => {
    map.on(eventType, e => {
      const { target, point } = e;
      const interactionSpec = getInteractionOnEvent({
        eventType,
        map: target,
        point,
        interactions,
      });

      if (!interactionSpec) return;

      const { interaction, feature, layerId } = interactionSpec;

      callback({ event: e, map, layerId, feature, interaction, eventType });
    });
  });

  /**
   * Mouseleave events for mouseover triggers
   */
  interactions.forEach(interaction => {
    const { id, trigger } = interaction;
    if (trigger !== 'mouseover') return;

    const eventType = 'mouseleave';
    map.on(eventType, id, event => {
      callback({ event, map, layerId: id, interaction, eventType });
    });
  });

  /**
   *  Display a pointer cursor over click zones
   */
  map.on('mousemove', e => {
    const { target, point } = e;
    const interactionSpec = getInteractionOnEvent({
      eventType: 'click',
      map: target,
      point,
      interactions,
    });

    const canvas = target.getCanvas();
    if (interactionSpec) {
      canvas.style.cursor = 'pointer';
    } else {
      canvas.style.cursor = '';
    }
  });
}

export const checkContraints = ({
  map,
  constraints = [],
  feature: { properties: { cluster } = {} } = {},
}) => {
  if (!constraints.length) return true;
  const currentZoom = map.getZoom();
  return constraints.reduce((prev, {
    minZoom = 0,
    maxZoom = Infinity,
    withLayers = [],
    isCluster,
  }) => {
    const checkZoom = minZoom !== undefined && maxZoom !== undefined
      ? currentZoom <= maxZoom && currentZoom >= minZoom
      : true;

    const checkLayers = withLayers.reduce((prevCheck, layer) => {
      const match = layer.match(/^(!)?(.+)$/);

      const visible = match[1] !== '!';
      const layerId = match[2];

      return prevCheck
        && visible
        ? (map.getLayer(layerId) !== undefined
          && map.getLayoutProperty(layerId, 'visibility') === 'visible')
        : (!map.getLayer(layerId)
          || map.getLayoutProperty(layerId, 'visibility') === 'none');
    }, true);
    const checkCluster = isCluster === undefined
      || cluster === isCluster;

    return prev || (checkZoom && checkLayers && checkCluster);
  }, false);
};

export default {
  toggleLayerVisibility,
  getOpacityProperty,
  setLayerOpacity,
  getInteractionOnEvent,
  setInteractions,
  checkContraints,
};
