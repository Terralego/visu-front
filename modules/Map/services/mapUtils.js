import bbox from '@turf/bbox';

import { PREFIXES } from './cluster';

const PREV_STATE = {};

export const getLayers = (map, layerId) => {
  const regexp = new RegExp(`^${layerId}(-(${PREFIXES.join('|')}))?(-[0-9]+)?$`);
  return map.getStyle().layers
    .filter(({ id }) => id.match(regexp));
};

export function toggleLayerVisibility (map, layerId, visibility) {
  getLayers(map, layerId)
    .forEach(({ id }) => map.setLayoutProperty(id, 'visibility', visibility));
}

export function getOpacityProperty (type) {
  switch (type) {
    case 'background':
      return 'background-opacity';
    case 'fill':
      return 'fill-opacity';
    case 'line':
      return 'line-opacity';
    case 'symbol':
      return 'icon-opacity';
    case 'raster':
      return 'raster-opacity';
    case 'circle':
      return 'circle-opacity';
    case 'fill-extrusion':
      return 'fill-extrusion-opacity';
    default:
      return null;
  }
}

export function setLayerOpacity (map, layerId, opacity) {
  getLayers(map, layerId)
    .forEach(layer => {
      const property = getOpacityProperty(layer.type);
      if (property) {
        map.setPaintProperty(layer.id, property, opacity);
      }
    });
}

export const checkContraints = ({
  map,
  constraints = [],
  feature: { properties: { cluster } = {} } = {},
}) => {
  if (!constraints.length) return true;
  const currentZoom = map.getZoom();
  return constraints.reduce((prev, {
    minZoom = 0,
    maxZoom = Infinity,
    withLayers = [],
    isCluster,
  }) => {
    const checkZoom = currentZoom <= maxZoom && currentZoom >= minZoom;

    const checkLayers = withLayers.reduce((prevCheck, layer) => {
      const match = layer.match(/^(!)?(.+)$/);

      const visible = match[1] !== '!';
      const layerId = match[2];
      return prevCheck && visible
        ? getLayers(map, layerId).reduce((subCheck, { id }) =>
          (subCheck ||
          map.getLayoutProperty(id, 'visibility') === 'visible'), false)
        : getLayers(map, layerId).reduce((subCheck, { id }) =>
          (subCheck &&
          map.getLayoutProperty(id, 'visibility') === 'none'), true);
    }, true);

    const checkCluster = isCluster === undefined
      || !!cluster === !!isCluster;

    return prev || (checkZoom && checkLayers && checkCluster);
  }, false);
};

export function getInteractionsOnEvent ({
  eventType,
  map,
  point,
  interactions: eventInteractions,
}) {
  const features = map.queryRenderedFeatures(point);

  let interactions = false;

  features.some(feature => {
    const { layer: { id: layerId } } = feature;

    const foundInteractions = eventInteractions.filter(({ id, trigger = 'click', constraints }) => {
      const found = getLayers(map, id).find(({ id: compatibleId }) => layerId === compatibleId);

      if (!found || (constraints && !checkContraints({ map, constraints, feature }))) {
        return false;
      }

      return found &&
        eventType === (trigger === 'mouseover' ? 'mousemove' : trigger);
    });

    if (!foundInteractions.length) return false;

    interactions = {
      interactions: foundInteractions,
      feature,
      layerId,
    };

    return true;
  });

  return interactions;
}

const listenerWaiter = {};
export function setInteractions ({ map, interactions, callback }) {
  const eventsTypes = new Set(interactions.reduce((triggers, { trigger = 'click' }) => [...triggers, trigger], []));

  if (eventsTypes.has('mouseover')) {
    eventsTypes.add('mousemove');
    eventsTypes.delete('mouseover');
  }

  /**
   * Mouseleave events for mouseover triggers
   * /!\ this listener MUST be before the mousemove
   */
  interactions.forEach(interaction => {
    const { id, trigger } = interaction;
    if (trigger !== 'mouseover') return;

    const eventType = 'mouseleave';
    map.on(eventType, id, event => {
      const features = map.queryRenderedFeatures(PREV_STATE.point);
      const feature = features.find(({ layer: { id: layerId } }) => id === layerId);
      callback({ event, map, layerId: id, interaction, feature, eventType });
    });
  });

  const listener = (e, eventType) => {
    const { target, point } = e;
    if (eventType === 'mousemove') {
      PREV_STATE.point = point;
    }
    const interactionsSpec = getInteractionsOnEvent({
      eventType,
      map: target,
      point,
      interactions,
    });

    if (!interactionsSpec) return;

    const { interactions: filteredInteractionsSpec, feature, layerId } = interactionsSpec;
    filteredInteractionsSpec.forEach(interaction =>
      callback({ event: e, map, layerId, feature, interaction, eventType }));
  };
  eventsTypes.forEach(eventType => {
    map.on(eventType, e => {
      clearTimeout(listenerWaiter[eventType]);
      listenerWaiter[eventType] = setTimeout(() => listener(e, eventType), eventType === 'mousemove' ? 200 : 100);
    });
  });

  /**
   *  Display a pointer cursor over click zones
   */
  map.on('mousemove', e => {
    const { target, point } = e;
    const interactionsSpec = getInteractionsOnEvent({
      eventType: 'click',
      map: target,
      point,
      interactions,
    });

    const canvas = target.getCanvas();
    if (interactionsSpec) {
      canvas.style.cursor = 'pointer';
    } else {
      canvas.style.cursor = '';
    }
  });
}

export function fitZoom ({ feature, map }) {
  map.fitBounds(bbox({ type: 'FeatureCollection', features: [feature] }));
}

export default {
  toggleLayerVisibility,
  getOpacityProperty,
  setLayerOpacity,
  getInteractionsOnEvent,
  setInteractions,
  checkContraints,
  fitZoom,
};
