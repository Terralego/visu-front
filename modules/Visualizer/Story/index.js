import Story from './Story';
//import { connectView } from '../context';

//export default connectView('map', 'setLegends')(Story);
export default Story;
