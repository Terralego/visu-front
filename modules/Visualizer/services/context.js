import React from 'react';

import connect from '../../../utils/connect';

const context = React.createContext();
export const { Provider, Consumer } = context;

export const connectVisualizerProvider = connect(context);

export default context;
