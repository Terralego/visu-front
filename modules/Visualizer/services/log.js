import debug from 'debug';

export const log = debug('terralego:Visualizer');

export default log;
