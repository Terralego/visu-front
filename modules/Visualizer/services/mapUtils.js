export function toggleLayerVisibility (map, layerId, visibility) {
  map.setLayoutProperty(layerId, 'visibility', visibility);
}

export function getOpacityProperty (type) {
  switch (type) {
    case 'background':
      return 'background-opacity';
    case 'fill':
      return 'fill-opacity';
    case 'line':
      return 'line-opacity';
    case 'symbol':
      return 'icon-opacity';
    case 'raster':
      return 'raster-opacity';
    case 'circle':
      return 'circle-opacity';
    case 'fill-extrusion':
      return 'fill-extrusion-opacity';
    default:
      return null;
  }
}

export function setLayerOpacity (map, layerId, opacity) {
  const layer = map.getLayer(layerId);
  const property = getOpacityProperty(layer.type);
  if (property) {
    map.setPaintProperty(layerId, property, opacity);
  }
}

export function getInteractionOnEvent ({ eventType, map, point, interactions }) {
  const features = map.queryRenderedFeatures(point);
  let interaction = false;

  features.some(feature => {
    const { layer: { id: layerId } } = feature;

    const foundInteraction = interactions.find(({ id, trigger = 'click' }) => {
      const found = id === layerId;

      return found &&
        eventType === (trigger === 'mouseover' ? 'mousemove' : trigger);
    });

    if (!foundInteraction) return false;

    interaction = {
      interaction: foundInteraction,
      feature,
      layerId,
    };

    return true;
  });

  return interaction;
}

export function setInteractions ({ map, interactions, callback }) {
  const eventsTypes = new Set(interactions.reduce((triggers, { trigger = 'click' }) => [...triggers, trigger], []));

  if (eventsTypes.has('mouseover')) {
    eventsTypes.add('mousemove');
    eventsTypes.delete('mouseover');
  }

  eventsTypes.forEach(eventType => {
    map.on(eventType, e => {
      const { target, point } = e;
      const interactionSpec = getInteractionOnEvent({
        eventType,
        map: target,
        point,
        interactions,
      });

      if (!interactionSpec) return;

      const { interaction, feature, layerId } = interactionSpec;

      callback({ event: e, map, layerId, feature, interaction, eventType });
    });
  });

  /**
   * Mouseleave events for mouseover triggers
   */
  interactions.forEach(interaction => {
    const { id, trigger } = interaction;
    if (trigger !== 'mouseover') return;

    const eventType = 'mouseleave';
    map.on(eventType, id, event => {
      callback({ event, map, layerId: id, interaction, eventType });
    });
  });

  /**
   *  Display a pointer cursor over click zones
   */
  map.on('mousemove', e => {
    const { target, point } = e;
    const interactionSpec = getInteractionOnEvent({
      eventType: 'click',
      map: target,
      point,
      interactions,
    });

    const canvas = target.getCanvas();
    if (interactionSpec) {
      canvas.style.cursor = 'pointer';
    } else {
      canvas.style.cursor = '';
    }
  });
}

export default {
  toggleLayerVisibility,
  getOpacityProperty,
  setLayerOpacity,
  getInteractionOnEvent,
  setInteractions,
};
