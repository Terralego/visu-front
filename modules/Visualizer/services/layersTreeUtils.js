import { PREFIX_SOURCE } from '../../Map/services/cluster';

export const INITIAL_FILTERS = new Map();

export const isCluster = (source, layerId) => !!source.match(new RegExp(`^${layerId}-${PREFIX_SOURCE}-[0-9]+`));

export function initLayersStateAction (layersTree) {
  const layersTreeState = new Map();
  function reduceLayers (group, map) {
    return group.reduce((layersStateMap, layer) => {
      const { initialState = {}, sublayers } = layer;
      const { active } = initialState;
      if (sublayers) {
        initialState.sublayers = initialState.sublayers || sublayers.map((_, k) =>
          (k === 0 && !!active));
      }
      if (layer.group) {
        return reduceLayers(layer.layers, layersStateMap);
      }
      initialState.opacity = initialState.opacity === undefined
        ? 1
        : initialState.opacity;
      layersStateMap.set(layer, {
        active: false,
        opacity: 1,
        ...initialState,
      });
      return layersStateMap;
    }, map);
  }
  return reduceLayers(layersTree, layersTreeState);
}

export function setLayerStateAction (layer, layerState, prevLayersTreeState) {
  const layersTreeState = new Map(prevLayersTreeState);
  const prevLayerState = layersTreeState.get(layer);
  const newLayerState = { ...layerState };

  if (!prevLayerState) return prevLayersTreeState;

  if (prevLayerState.sublayers && !prevLayerState.sublayers.find(sl => sl)) {
    newLayerState.sublayers = [...prevLayerState.sublayers];
    newLayerState.sublayers[0] = true;
  }
  if (newLayerState.table) {
    // Easiest to to read as transform Map in Array and run a .map() on it
    // eslint-disable-next-line no-param-reassign
    layersTreeState.forEach(layState => { layState.table = false; });
  }
  layersTreeState.set(layer, {
    ...prevLayerState,
    ...newLayerState,
  });
  return layersTreeState;
}

export function layersTreeStatesHaveChanged (layersTreeState, prevLayersTreeState, fields = ['active']) {
  return !fields.reduce((all, field) =>
    all && Array.from(layersTreeState).reduce((suball, [layer, state = {}]) =>
      suball && state[field] === (prevLayersTreeState.get(layer) || {})[field],
    all),
  true);
}

export function selectSublayerAction (layer, sublayer, prevLayersTreeState) {
  const layersTreeState = new Map(prevLayersTreeState);
  const layerState = layersTreeState.get(layer);
  layerState.sublayers = layerState.sublayers.map((_, k) => k === sublayer);
  layersTreeState.set(layer, { ...layerState });
  return layersTreeState;
}

export const filterLayersStatesFromLayersState = (
  layersTreeState = [],
  check = ({ active }) => !!active,
) => Array
  .from(layersTreeState)
  .reduce((allLayers, [layer, state]) =>
    ((check(state)) ? [...allLayers, [layer, state]] : allLayers), []);

export const filterLayersFromLayersState = (layersTreeState, check) =>
  filterLayersStatesFromLayersState(layersTreeState, check)
    .filter(([{ filters: { layer } = {} }]) => layer)
    .map(([{ filters: { layer } }]) => layer);

export const hasTable = layersTreeState =>
  filterLayersFromLayersState(layersTreeState, ({ table }) => table).length > 0;

export const hasWidget = layersTreeState =>
  filterLayersStatesFromLayersState(layersTreeState, ({ widgets = [] }) => widgets.length > 0)
    .length > 0;

export const filterFeatures = (
  map,
  features/* = [ { layer: String, features: [id1, id2, â€¦] } ] */,
  layersTreeState,
) => {
  Array.from(layersTreeState).forEach(([{
    layers, filters: { layer } = {},
  }, {
    active,
  }]) => {
    if (!active || !layers) return;
    const layerFeatures = features
      .find(({ layer: fLayer }) => fLayer === layer);
    const { features: ids = [] } = layerFeatures || {};

    const filter = ids.length ? ['match', ['get', '_id'], ids, true, false] : null;
    layers.forEach(layerId => {
      const paintLayer = map.getLayer(layerId);
      if (!paintLayer) return;

      if (isCluster(paintLayer.source, layerId)) {
        // Force to upgrade cluster data
        map.fire('refreshCluster');
        return;
      }

      if (!INITIAL_FILTERS.has(layerId)) {
        INITIAL_FILTERS.set(layerId, map.getFilter(layerId));
      }
      map.setFilter(layerId, layerFeatures ? filter : INITIAL_FILTERS.get(layerId));
    });
  });
};

export const resetFilters = (map, layersTreeState) => {
  Array.from(layersTreeState).forEach(([{
    layers = [], filters,
  }]) => {
    if (!filters) return;

    layers.forEach(layerId => {
      const paintLayer = map.getLayer(layerId);

      if (!map.getLayer(layerId)) return;

      if (isCluster(paintLayer.source, layerId)) {
        // Force to upgrade cluster data
        map.fire('refreshCluster');
        return;
      }
      if (!INITIAL_FILTERS.has(layerId)) return;
      map.setFilter(layerId, INITIAL_FILTERS.get(layerId));
    });
  });
};

export default {
  initLayersStateAction,
  selectSublayerAction,
  setLayerStateAction,
  filterLayersFromLayersState,
  hasTable,
  hasWidget,
  filterFeatures,
  resetFilters,
};
