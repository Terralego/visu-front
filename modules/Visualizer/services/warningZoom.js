import { getRelatedLayers } from '../../Map/services/mapUtils';

export function displayWarningAccordingToZoom (map, layer) {
  if (!map) return {};
  const currentZoom = map.getZoom();
  const layers = layer.layers.reduce((prev, layerId) =>
    [...prev, ...getRelatedLayers(map, layerId)], []);
  const layerZoom = layers
    .reduce(({ minzoom: prevMinzoom, maxzoom: prevMaxzoom }, { id }) => {
      const { minzoom = 0, maxzoom = 24 } = map.getLayer(id) || {};
      return {
        minzoom: Math.min(minzoom, prevMinzoom),
        maxzoom: Math.max(maxzoom, prevMaxzoom),
      };
    }, { minzoom: 24, maxzoom: 0 });
  const isNotDisplayed = currentZoom < layerZoom.minzoom || currentZoom > layerZoom.maxzoom;

  return { display: isNotDisplayed, minZoomLayer: layerZoom.minzoom };
}

export default { displayWarningAccordingToZoom };
