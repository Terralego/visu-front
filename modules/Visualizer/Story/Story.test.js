import React from 'react';
import renderer from 'react-test-renderer';

import Story from './Story';
import { toggleLayerVisibility } from '../../Map/services/mapUtils';

jest.mock('../../Map/services/mapUtils', () => ({
  toggleLayerVisibility: jest.fn(),
}));

const story = {
  beforeEach: [{
    layers: [
      'foo-*',
      'bar',
    ],
    active: false,
  }],
  slides: [{
    title: 'Slide 1',
    content: `
  <p>
    Content 1
  </p>
`,
  }, {
    title: 'Slide 2',
    content: `
<p>
  Content 2
</p>
`,
    layouts: [{
      layers: ['foo-1a', 'foo-1b'],
      active: true,
    }],
  }, {
    title: 'Slide 3',
    content: `
<p>
  Content 3
</p>
`,
    layouts: [{
      layers: ['foo-2'],
      active: true,
    }],
    legends: [{
      title: 'Legend 1',
      items: [
        { label: 'a', color: '#ff8c00' },
        { label: 'b', color: '#965096' },
        { label: 'c', color: '#ff78a0' },
        { label: 'd', color: '#003c82' },
        { label: 'e', color: '#f5000a' },
        { label: 'f', color: '#c8be00' },
        { label: 'g', color: '#90ff77' },
        { label: 'h', color: '#007dff' },
        { label: 'i', color: '#1e6414' },
      ],
    }],
  }, {
    title: 'Slide 4',
    content: `
<p>
  Content 4
</p>
    `,
    layouts: [{
      layers: ['foo-3'],
      active: true,
    }],
    legends: [{
      title: 'Legend 2',
      items: [
        { label: '1', color: '#b91a1c' },
        { label: '2', color: '#b3de69' },
        { label: '3', color: '#1f78b4' },
        { label: '4', color: '#a6a5ae' },
        { label: '5', color: '#ec7c30' },
        { label: '6', color: '#fedb2a' },
        { label: '7', color: '#8fa9db' },
        { label: '8', color: '#9e480d' },
      ],
    }],
  }, {
    title: 'Slide 5',
    content: `
<p>
  Content 5
</p>
`,
    layouts: [{
      layers: ['foo-4'],
      active: true,
    }],
    legends: [{
      title: 'Legend 3',
      items: [
        { label: 'a', color: '#fcbba1', shape: 'circle', radius: 5 },
        { label: 'aa', color: '#fc9272', shape: 'circle', radius: 10 },
        { label: 'aaa', color: '#fb6a4a', shape: 'circle', radius: 15 },
        { label: 'aaaa', color: '#ef3b2c', shape: 'circle', radius: 20 },
        { label: 'aaaaa', color: '#cb181d', shape: 'circle', radius: 25 },
        { label: 'aaaaaa', color: '#99000d', shape: 'circle', radius: 30 },
      ],
    }],
  }, {
    title: 'Slide 6',
    content: `
<p>
  Content 6
</p>
`,
    layouts: [{
      layers: ['bar'],
      active: true,
    }],
  }],
};

it('should render', () => {
  const tree = renderer.create(
    <Story
      story={story}
    />,
  );
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should render a step', () => {
  const tree = renderer.create(
    <Story
      story={story}
    />,
  );
  tree.getInstance().setState({ step: 1 });
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should render last step', () => {
  const tree = renderer.create(
    <Story
      story={story}
    />,
  );
  tree.getInstance().setState({ step: 5 });
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should init map on mount and update', () => {
  const instance = new Story({});
  instance.initMap = jest.fn();
  instance.displayStep = () => null;

  instance.componentDidUpdate({ map: {} }, {});
  expect(instance.initMap).toHaveBeenCalled();
  instance.initMap.mockClear();

  instance.componentDidUpdate({}, {});
  expect(instance.initMap).not.toHaveBeenCalled();
});

it('should update step', () => {
  const instance = new Story({});
  instance.displayStep = jest.fn();

  instance.componentDidUpdate({ map: {} }, {});
  expect(instance.displayStep).toHaveBeenCalled();
  instance.displayStep.mockClear();

  instance.props.map = {};
  instance.componentDidUpdate({ map: instance.props.map }, { step: 1 });
  expect(instance.displayStep).toHaveBeenCalled();
  instance.displayStep.mockClear();

  instance.state.step = 1;
  instance.componentDidUpdate({ map: instance.props.map }, { step: 1 });
  expect(instance.displayStep).not.toHaveBeenCalled();
});

it('should set prev step', () => {
  const instance = new Story({
    story,
  });
  let callback;
  instance.setState = jest.fn(fn => { callback = fn; });
  instance.prevStep();
  expect(instance.setState).toHaveBeenCalledWith(expect.any(Function));

  expect(callback({ step: 5 })).toEqual({ step: 4 });
  expect(callback({ step: 4 })).toEqual({ step: 3 });
  expect(callback({ step: 3 })).toEqual({ step: 2 });
  expect(callback({ step: 2 })).toEqual({ step: 1 });
  expect(callback({ step: 1 })).toEqual({ step: 0 });
  expect(callback({ step: 0 })).toEqual({ step: 5 });
});

it('should set next step', () => {
  const instance = new Story({
    story,
  });
  let callback;
  instance.setState = jest.fn(fn => { callback = fn; });
  instance.nextStep();
  expect(instance.setState).toHaveBeenCalledWith(expect.any(Function));

  expect(callback({ step: 5 })).toEqual({ step: 0 });
  expect(callback({ step: 4 })).toEqual({ step: 5 });
  expect(callback({ step: 3 })).toEqual({ step: 4 });
  expect(callback({ step: 2 })).toEqual({ step: 3 });
  expect(callback({ step: 1 })).toEqual({ step: 2 });
  expect(callback({ step: 0 })).toEqual({ step: 1 });
});

it('should update map', () => {
  const instance = new Story({});
  instance.displayStep = jest.fn();
  instance.initMap();
  expect(instance.displayStep).not.toHaveBeenCalled();

  instance.props.map = {};

  instance.initMap();
  expect(instance.displayStep).toHaveBeenCalled();
});

it('should display step', () => {
  const setLegends = jest.fn();
  const map = {
    getStyle: () => ({
      layers: [{
        id: 'foo-1a',
      }, {
        id: 'foo-1b',
      }, {
        id: 'foo-2',
      }, {
        id: 'foo-3',
      }, {
        id: 'foo-4',
      }, {
        id: 'bar',
      }, {
        id: 'bapapa',
      }],
    }),
  };
  const instance = new Story({ story, setLegends });
  instance.displayStep();
  expect(toggleLayerVisibility).not.toHaveBeenCalled();

  instance.props.map = map;
  instance.displayStep();
  expect(toggleLayerVisibility).toHaveBeenCalledTimes(1);
  expect(setLegends).toHaveBeenCalled();
  toggleLayerVisibility.mockClear();

  instance.state.step = 1;
  instance.displayStep();
  expect(toggleLayerVisibility).toHaveBeenCalledTimes(3);
  toggleLayerVisibility.mockClear();

  instance.props.story.beforeEach = [{}];
  instance.displayStep();
  expect(toggleLayerVisibility).toHaveBeenCalledTimes(2);
  toggleLayerVisibility.mockClear();

  instance.props.story.beforeEach = undefined;
  instance.displayStep();
  expect(toggleLayerVisibility).toHaveBeenCalledTimes(2);
  toggleLayerVisibility.mockClear();
});
