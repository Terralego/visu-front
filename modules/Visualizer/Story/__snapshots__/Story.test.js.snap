// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  className="storytelling"
>
  <div
    className="storytelling__content"
  >
    <h2>
      Slide 1
    </h2>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "
  <p>
    Content 1
  </p>
",
        }
      }
    />
  </div>
  <div
    className="storytelling__buttons"
  >
    <button
      className="bp3-button bp3-large bp3-intent-primary"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      type="button"
    >
      <span
        className="bp3-button-text"
      >
        Démarrer
      </span>
      <span
        className="bp3-icon bp3-icon-chevron-right"
        icon="chevron-right"
      >
        <svg
          data-icon="chevron-right"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <desc>
            chevron-right
          </desc>
          <path
            d="M10.71 7.29l-4-4a1.003 1.003 0 00-1.42 1.42L8.59 8 5.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4-4c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71z"
            fillRule="evenodd"
          />
        </svg>
      </span>
    </button>
  </div>
   
</div>
`;

exports[`should render a step 1`] = `
<div
  className="storytelling"
>
  <div
    className="storytelling__content"
  >
    <h2>
      Slide 2
    </h2>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "
<p>
  Content 2
</p>
",
        }
      }
    />
  </div>
  <div
    className="storytelling__buttons storytelling__buttons--justified"
  >
    <button
      className="bp3-button bp3-large"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      type="button"
    >
      <span
        className="bp3-icon bp3-icon-chevron-left"
        icon="chevron-left"
      >
        <svg
          data-icon="chevron-left"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <desc>
            chevron-left
          </desc>
          <path
            d="M7.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71l-4 4C5.11 7.47 5 7.72 5 8c0 .28.11.53.29.71l4 4a1.003 1.003 0 001.42-1.42L7.41 8z"
            fillRule="evenodd"
          />
        </svg>
      </span>
      <span
        className="bp3-button-text"
      >
        Précédent
      </span>
    </button>
    <button
      className="bp3-button bp3-large bp3-intent-primary"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      type="button"
    >
      <span
        className="bp3-button-text"
      >
        Suivant
      </span>
      <span
        className="bp3-icon bp3-icon-chevron-right"
        icon="chevron-right"
      >
        <svg
          data-icon="chevron-right"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <desc>
            chevron-right
          </desc>
          <path
            d="M10.71 7.29l-4-4a1.003 1.003 0 00-1.42 1.42L8.59 8 5.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4-4c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71z"
            fillRule="evenodd"
          />
        </svg>
      </span>
    </button>
  </div>
   
</div>
`;

exports[`should render last step 1`] = `
<div
  className="storytelling"
>
  <div
    className="storytelling__content"
  >
    <h2>
      Slide 6
    </h2>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "
<p>
  Content 6
</p>
",
        }
      }
    />
  </div>
  <div
    className="storytelling__buttons storytelling__buttons--justified"
  >
    <button
      className="bp3-button bp3-large"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      type="button"
    >
      <span
        className="bp3-icon bp3-icon-chevron-left"
        icon="chevron-left"
      >
        <svg
          data-icon="chevron-left"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <desc>
            chevron-left
          </desc>
          <path
            d="M7.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71l-4 4C5.11 7.47 5 7.72 5 8c0 .28.11.53.29.71l4 4a1.003 1.003 0 001.42-1.42L7.41 8z"
            fillRule="evenodd"
          />
        </svg>
      </span>
      <span
        className="bp3-button-text"
      >
        Précédent
      </span>
    </button>
    <button
      className="bp3-button bp3-large bp3-intent-primary"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      type="button"
    >
      <span
        className="bp3-icon bp3-icon-step-backward"
        icon="step-backward"
      >
        <svg
          data-icon="step-backward"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <desc>
            step-backward
          </desc>
          <path
            d="M12 3c-.24 0-.44.09-.62.23l-.01-.01L7 6.72V4c0-.55-.45-1-1-1H5c-.55 0-1 .45-1 1v8c0 .55.45 1 1 1h1c.55 0 1-.45 1-1V9.28l4.38 3.5.01-.01c.17.14.37.23.61.23.55 0 1-.45 1-1V4c0-.55-.45-1-1-1z"
            fillRule="evenodd"
          />
        </svg>
      </span>
      <span
        className="bp3-button-text"
      >
        Recommencer
      </span>
    </button>
  </div>
   
</div>
`;
